contracts/upgradeable_contracts/modules/interest/AAVEInterestERC20.sol

```diff
@@ -0,0 +1,240 @@
+     * Only Omnibridge contract is allowed to call this method.
+     * Converts aTOKENs into _amount of TOKENs.
+     * @param _token address of the invested token contract.
+     * @param _amount minimal amount of tokens to withdraw.
+     */
+    function withdraw(address _token, uint256 _amount) external override onlyMediator {
+        InterestParams storage params = interestParams[_token];
+        uint256 invested = params.investedAmount;
+        uint256 redeemed = _safeWithdraw(_token, _amount > invested ? invested : _amount);
+        params.investedAmount = redeemed > invested ? 0 : invested - redeemed;
+        IERC20(_token).transfer(mediator, redeemed);
+    }
+
+    /**
+     * @dev Tells the current accumulated interest on the invested tokens, that can be withdrawn and payed to the interest receiver.
+     * @param _token address of the invested token contract.
+     * @return amount of accumulated interest.
+     */
+    function interestAmount(address _token) public view returns (uint256) {
+        InterestParams storage params = interestParams[_token];
+        (IAToken aToken, uint96 dust) = (params.aToken, params.dust);
+        uint256 balance = aToken.balanceOf(address(this));
+        // small portion of tokens are reserved for possible truncation/round errors
+        uint256 reserved = params.investedAmount.add(dust);
+        return balance > reserved ? balance - reserved : 0;
+    }
+
+    /**
+     * @dev Pays collected interest for the underlying token.
+     * Anyone can call this function.
+     * Earned interest is withdrawn and transferred to the specified interest receiver account.
+     * @param _token address of the invested token contract in which interest should be paid.
+     */
+    function payInterest(address _token) external {
+        InterestParams storage params = interestParams[_token];
+        uint256 interest = interestAmount(_token);
+        require(interest >= params.minInterestPaid);
+        _transferInterest(params.interestReceiver, address(_token), _safeWithdraw(_token, interest));
+    }
+
+    /**
+     * @dev Last-resort function for returning assets to the Omnibridge contract in case of some failures in the logic.
+     * Disables this contract and transfers locked tokens back to the mediator.
+     * Only owner is allowed to call this method.
+     * @param _token address of the invested token contract that should be disabled.
+     */
+    function forceDisable(address _token) external onlyOwner {
+        InterestParams storage params = interestParams[_token];
+        IAToken aToken = params.aToken;
+
+        uint256 aTokenBalance = aToken.balanceOf(address(this));
+        // try to redeem all aTokens
+        try lendingPool().withdraw(_token, aTokenBalance, mediator) {
+            uint256 balance = IERC20(_token).balanceOf(address(this));
+            IERC20(_token).transfer(mediator, balance);
+        } catch {
+            aToken.transfer(mediator, aTokenBalance);
```

Comment:
And what will be done after that? Do we expect an upgrade of the OB contracts if this situation happens? Is it assumed that the issue could be addressed by `fixMediatorBalance`?

Reply:
Yes, I assume that if some contract logic is broken, we will need to call this function in order to prevent tokens loss. It should not be called in normal operation conditions. But if it is called anyway, and tokens withdrawal is not possible for some reason (e.g. insufficient liquidity), then I would expect an additional required upgrade in the OB contract itself. 

This function also looks useful, since interest implementation contracts are non-upgradeable. As an alternative solution, we can make interest implementations upgradeable, but I don't know if it will make maintenance easier.